prereise/gather/griddata/hifld/data_process/generators.py

```diff
@@ -22,54 +26,131 @@ def floatify(value, default=float("nan")):
-    substation within that ZIP code, 4) if only ZIP code is available (no location)
-    but no substations exist in that ZIP code, return NA.
-
-    :param pandas.Series generator: one generating unit from data frame.
-    :param pandas.GroupBy substation_groupby: data frame of substations, grouped by
-        (interconnect, ZIP).
-    :return: (*int/pd.NA*) -- substation ID if the generator can be mapped successfully
-        to a substation, else pd.NA.
+def map_generators_to_sub_by_location(
+    generators, substations, inplace=True, report_worst=None
+):
+    """Determine the closest substation to each generator. For generators without
+    latitude and longitude, an attempt will be made to match via ZIP code, and failing
+    that a pandas.NA value will be returned.
+
+    :param pandas.DataFrame generators: generator data. Required columns:
+        'interconnect', 'lat', 'lon', 'state', 'ZIP'.
+    :param pandas.DataFrame substations: substation data. Required columns:
+        'interconnect', 'lat', 'lon', 'STATE', 'ZIP'.
+    :param bool inplace: whether to modify the generator table inplace with a new
+        'sub_id' column or to return a new one. If ``inplace`` is `True`, entries in
+        `generators` which have non-sensical combinations of 'state' and 'interconnect'
+        columns will have their 'interconnect' entries modified.
+    :param int report_worst: if not None, display the distances of the worst N mappings.
+    :return: (*pandas.DataFrame/None*) -- if ``inplace`` is `False`, return the modified
+        DataFrame; otherwise return nothing.
     """
-    lookup_params = tuple(generator.loc[["interconnect", "ZIP"]])
-    if pd.isna(generator["lat"]) or pd.isna(generator["lon"]):
-        # No location available
-        try:
-            matching_subs = substation_groupby.get_group(lookup_params)
-            return matching_subs.index[0]
-        except KeyError:
-            return pd.NA
-    try:
-        # This ZIP code contains substations, this block will execute successfully
-        matching_subs = substation_groupby.get_group(lookup_params)
-    except KeyError:
-        # If this ZIP code does not contain substations, this block will execute, and
-        # we select a set of 'nearby' substations
-        zip_range = [int(generator.loc["ZIP"]) + offset for offset in range(-100, 101)]
-        zip_range_strings = [str(z).rjust(5, "0") for z in zip_range]
-        try:
-            matching_subs = pd.concat(
-                [
-                    substation_groupby.get_group((generator.loc["interconnect"], z))
-                    for z in zip_range_strings
-                    if (generator.loc["interconnect"], z) in substation_groupby.groups
-                ]
-            )
-        except ValueError:
-            # If no matching subs within the given interconnection and ZIPs, give up
+
+    def get_closest_substation(generator, state_trees, subs_state_lookup):
```

Comment:
Shall we rename the input variables into `voltage_trees` and `subs_voltage_lookup` instead?

Reply:
Good catch. As it was written, those parameter were never being used, and the same-name versions from the parent scope were being used instead (which still behaved identically, but this function call signature didn't really make sense).

Reply:
Done.